File: login/javascript/signup.js
/*
 *	@Project  : Helios
 *	@Architecture : Kim Bom
 *	signup.js
 *
 *	@Created by KimBom On 2016. 1. 2...
 *	@Copyright (C) 2015 KimBom. All rights reserved.
 */

/**
 * @name CheckBlank
 * @check1 : 공백검사
 * @check2 : ID,PW,EMAIL,CPW,FILE 유효성 검사
 */
function CheckSubmit() {
	var id = document.getElementById("ID").value;
	var pw = document.getElementById("PW").value;
	var name = document.getElementById("NAME").value;
	var guild = document.getElementById("GUILD").value;
	var phone = document.getElementById("PHONE").value;
	var email = document.getElementById("EMAIL").value;
	var file = document.getElementById("FILE").value;
	// 공백검사
	if (id == "" || pw == "" || name == "" || guild == "" || phone == ""
			|| email == "" || file == "") {
		alert("Please enter all information.");
		return false;
	}
	// 아이디 검사
	if (id_keyup() == false) {
		alert("id error");
		return false;
	}
	// 비밀번호 검사
	if (pw_keyup() == false) {
		alert("password error");
		return false;
	}
	// 이메일 검사
	if (em_keyup() == false) {
		alert("email error");
		return false;
	}
	// 확인 비밀번호 검사
	if (cpw_keyup() == false) {
		alert("confirm password error");
		return false;
	}
	//확인 아이디 검사
	if(cid_keyup()==false){
		alert("confirm id error");
		return false;
	}
	// 파일 확장자 검사
	if (file_chk() == false) {
		alert("file upload error");
		return false;
	}
	// SHA3 암호화
	var salt=CryptoJS.lib.WordArray.random(32);
	var enc_pw=CryptoJS.SHA3(salt+pw);
	//var enc_id=CryptoJS.SHA3(id);
	document.getElementById("HPW").value = enc_pw;
	document.getElementById("HID").value = id;
	document.getElementById("HSALT").value = salt;
	return true;
	
}
/**
 * @name : id_keyup 
 * @check1 : id length<4
 * @check2 : id already exist
 * @check3 : only use alpha,digit,under bar
 */
function id_keyup() {
	var i = document.getElementById("ID");
	// 알파벳,숫자,언더바가 아니면 삭제한다.
	if (i.value.length > 0) {
		i.value = i.value.replace(/[^(\d|a-z|A-Z|_)]+/g, '');
	}
	var ids = document.getElementById("IDS").value;
	var already_id = ids.split(',');
	var id = document.getElementById("ID").value;
	if (id.length >= 4) {
		//var enc_id=CryptoJS.SHA3(id);
		var char4 = document.getElementById("char4");
		char4.style.visibility = "hidden";
		for ( var i in already_id) {
			// 이미 존재하는 ID일경우 오류 출력
			if (already_id[i] == id) {
				var sameid = document.getElementById("sameid");
				sameid.style.visibility = "visible";
				return false;
			}
		}
		var sameid = document.getElementById("sameid");
		sameid.style.visibility = "hidden";
	} else {
		// ID의 길이가 4이하면 오류메세지 출력
		var sameid = document.getElementById("sameid");
		sameid.style.visibility = "hidden";
		var char4 = document.getElementById("char4");
		char4.style.visibility = "visible";
		return false;
	}
	return true;
}
/**
 * @name : cid_keyup
 * @check1 : 비밀번호와 같은지 검사
 */
function cid_keyup() {
	var id = document.getElementById("ID").value;
	var cid = document.getElementById("CID").value;
	if (cid == id) {
		var c_id = document.getElementById("cid");
		c_id.style.visibility = "hidden";
		return true;
	} else {
		var c_id = document.getElementById("cid");
		c_id.style.visibility = "visible";
		return false;
	}
}
/**
 * @name : pw_keyup
 * @check1 : pw length<4
 */
function pw_keyup() {
	var pw = document.getElementById("PW").value;
	var i = document.getElementById("PW");
	if (i.value.length > 0) {
		i.value = i.value.replace(/[^(\d|a-z|A-Z|_)]+/g, '');
	}
	if(pw!=i.value){
		var ivc = document.getElementById("ivchar");
		ivc.style.visibility = "visible";
		return false;
	}else{
		var ivc = document.getElementById("ivchar");
		ivc.style.visibility = "hidden";
	}
	// 길이가 4이하면 에러
	if (pw.length >= 4) {
		var pw4 = document.getElementById("pass4");
		pw4.style.visibility = "hidden";
		return true;
	} else {
		var pw4 = document.getElementById("pass4");
		pw4.style.visibility = "visible";
		return false;
	}
}
/**
 * @name : cpw_keyup
 * @check1 : 비밀번호와 같은지 검사
 */
function cpw_keyup() {
	var pw = document.getElementById("PW").value;
	var cpw = document.getElementById("CPW").value;
	if (cpw == pw) {
		var c_pw = document.getElementById("cpass");
		c_pw.style.visibility = "hidden";
		return true;
	} else {
		var c_pw = document.getElementById("cpass");
		c_pw.style.visibility = "visible";
		return false;
	}
}
/**
 * @name : not empty keyup
 * @check1 : 공백검사
 * @param1 : 속성 ID
 * @param2 : 경고 메세지 ID
 */
function nempty_keyup(param1, param2) {
	var value = document.getElementById(param1).value;
	if (value == "") {
		var bar = document.getElementById(param2);
		bar.style.visibility = "visible";
	} else {
		var bar = document.getElementById(param2);
		bar.style.visibility = "hidden";
	}
}
/**
 * @name : email_keyup
 * @check1 : email check
 */
function em_keyup() {
	var email = document.getElementById("EMAIL").value;
	var regex = /^([\w-]+(?:\.[\w-]+)*)@((?:[\w-]+\.)*\w[\w-]{0,66})\.([a-z]{2,6}(?:\.[a-z]{2})?)$/;
	if (regex.test(email) === false) {
		var em = document.getElementById("emchk");
		em.style.visibility = "visible";
		return false;
	} else {
		var em = document.getElementById("emchk");
		em.style.visibility = "hidden";
		return true;
	}
}
/**
 * @name : tel_keyup
 * @check1 : 숫자만 입력가능(다른문자가 들어오면 제거)
 */
function tel_keyup() {
	var i = document.getElementById("PHONE");
	if (i.value.length > 0) {
		i.value = i.value.replace(/[^\d]+/g, '');
	}
	if (i.value == "") {
		var em = document.getElementById("phonechk");
		em.style.visibility = "visible";
		return true;
	} else {
		var em = document.getElementById("phonechk");
		em.style.visibility = "hidden";
	}
	return true;
}
/**
 * @name : file check
 * @check1 : 확장자검사
 */
function file_chk() {
	var file = document.getElementById("FILE").value;
	var execs = new Array(".jpg", ".gif",  ".bmp", ".png",".jpeg");
	var bSubmitCheck = false;
	if (!file) {
		alert("please select your photo");
		return;
	}
	while (file.indexOf("\\") != -1) {
		file = file.substring(file.indexOf("\\") + 1);
	}
	ext = file.substring(file.indexOf(".")).toLowerCase();
	for (var i = 0; i < execs.length; i++) {
		if (execs[i] == ext) {
			bSubmitCheck = true;
			break;
		}
	}
	if (!bSubmitCheck) {
		alert("enable upload file.\n\n" + (execs.join("  "))
				+ "please select again");
		return false;
	}
	return true;
}
/**
 * @name : file size check
 * @check1 : file size < 4mb
 */
function filesizechk(objFile) {
	var nMaxSize = 4 * 1024 * 1024; // 4 MB
	var nFileSize = objFile.files[0].size;
	if (nFileSize > nMaxSize) {
		document.getElementById("FILE").value = "";
		document.getElementById("fakeFile").value = "";
		alert("Please upload a file less than 4MB");
	} else {
		document.getElementById("fakeFile").value = document
				.getElementById("FILE").value;
	}
}
/**
 * @name : go file
 * @comment : 파일을 클릭하게 합니다.
 */
function goFile() {
	document.getElementById("FILE").click();
}

File: login/signup.pl
#!/usr/bin/perl
#
#	@Project  : Helios
#	@Architecture : Kim Bom
#	signup.pl
#
#	@Created by KimBom On 2016. 1. 2...
#	@Copyright (C) 2015 KimBom. All rights reserved.
#
use strict;
use warnings;
use CGI;
use DBI;
use File::Copy;
use Digest::SHA3 qw(sha3_512_hex);
use Crypt::Salt;
require 'info.pl';

#==============================CGI 작업을 준비합니다.==============================
my $q = new CGI;
my $con = DBI->connect( GetDB(), GetID(), GetPW() );

#==============================Form의 값을 받아옵니다.==============================
my $id          = $q->param('HID');
my $pw          = $q->param('HPW');
my $name        = $q->param('NAME');
my $guild       = $q->param('GUILD');
my $phone       = $q->param('PHONE');
my $email       = $q->param('EMAIL');
my $upload_file = $q->param('FILE');
my $salt        = $q->param('HSALT');
#$id='root';$pw='root';$name='root';$guild='root';$phone='1234';$email='root';$salt='abvs';
my $ids = "";    #데이터베이스의 아이디 목록입니다.
#==============================회원가입을 완료했을때 호출됩니다..==============================
if ($id) {
	$upload_file =~ m<.+\.(\w+)?$>;    #확장자를 추출합니다.
	my $photo_path = "$id-0-.$1";
	copy( $upload_file, "photo/$photo_path" );	#파일을 저장합니다.
	my $p_salt=salt(32);
	$pw=sha3_512_hex($p_salt.$pw);
	my $query = "INSERT INTO userinfo values(\'$id\',\'$pw\',\'$name\',\'$guild\',\'$phone\',\'$email\',\'$photo_path\',\'0\',\'$salt\',\'$p_salt\')";
	$con->do($query);
	$con->disconnect();
	print $q->redirect('login.pl');
}
else {
	my $query = "SELECT ui_id FROM userinfo";
	my $state = $con->prepare($query);
	$state->execute();
	while ( my @row = $state->fetchrow_array ) {
		$ids = $ids . $row[0] . ",";
	}
	chop($ids);
	$state->finish();
	$con->disconnect();
}
#=======================================DEBUG=================
my $visible="hidden";
#==============================CGI 부분==============================
print $q->header( -charset => "UTF-8" );
print <<EOF
<head>
	<title>BlueCandle</title>
	<link rel="stylesheet" type="text/css" href="css/signup.css" />
	<script src="javascript/sha3.js"></script>
	<script src="javascript/pbkdf2.js"></script>
	<script src="javascript/signup.js" type="text/javascript"></script>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
</head>
<body>
<section class="back">
		<form action="login.pl" method="post">
			<input type="submit" id="login"  value="back to login"></input>
		</form>
</selection>
<div class="container">	
	<form action="signup.pl" method="post" ENCTYPE="multipart/form-data">
	<div class="outer"><div class="inner"><div class="centered-top"></div><div class="centered">
		<div class="centered_inner_right"></div>
		<div id="msg" class="centered_inner_right2">
			<div class='warn' id="sameid" style="visibility:$visible;top:177px">The same id exist</div>
			<div class='warn' id="char4" style="visibility:$visible;top:137px">Id must be at least 4 char</div>
			<div class='warn' id="cid" style="visibility:$visible;top:145px">not matching ID</div>
			
			<div class='warn' id="pass4" style="visibility:$visible;top:154px">Password is too short</div>
			<div class='warn' id="ivchar" style="visibility:$visible;top:114px">Invaild character</div>
			<div class='warn' id="cpass" style="visibility:$visible;top:124px">not matching password</div>
			<div class='warn' id="namechk" style="visibility:$visible;top:134px">input your name</div>
			<div class='warn' id="guildchk" style="visibility:$visible;top:144px">input your guild</div>
			<div class='warn' id="phonechk" style="visibility:$visible;top:154px">input your phone</div>
			<div class='warn' id="emchk" style="visibility:$visible;top:164px">This is not a valid e-mail</div>
		</div>
		<div class="centered_inner_left">
			<p class="leftP">Your best id</p>
			<p class="leftP">Confirm id</p>
			<p class="leftP">Safety password</p>
			<p class="leftP">Confirm password</p>
			<p class="leftP">What`s your name?</p>
			<p class="leftP">Where do you belong?</p>
			<p class="leftP">Phone number</p>
			<p class="leftP">Your most used e-mail</p>
			<p class="leftP">Your pretty face</p>
		</div>
		<section class="webdesigntuts-workshop">
			<input type="hidden" id="IDS"  value="$ids"></input>
			<input type="hidden" id="HPW" name="HPW" value=""></input>
			<input type="hidden" id="HID" name="HID" value=""></input>
			<input type="hidden" id="HSALT" name="HSALT" value=""></input>
			
			<input type="text" id="ID" autocomplete="off" name="ID" placeholder="ID" onkeyup="id_keyup()" onblur="id_keyup()" maxlength="64"></input>
			<input type="text" id="CID" autocomplete="off" name="CID" placeholder="Confirm ID" onkeyup="cid_keyup()" onblur="cid_keyup()" maxlength="64"></input>
			<input type="password" id="PW" autocomplete="off" name="PW" placeholder="PW" onkeyup="pw_keyup()" onblur="pw_keyup()" maxlength="64"></input>
			<input type="password" id="CPW" autocomplete="off" name="CPW" placeholder="confirm PW" onkeyup="cpw_keyup()" onblur="pw_keyup()" maxlength="64"></input>
			<input type="text" id="NAME" autocomplete="off" name="NAME" placeholder="NAME" maxlength="64" onkeyup="nempty_keyup('NAME','namechk')" onblur="nempty_keyup('NAME','namechk')"></input>
			<input type="text" id="GUILD" autocomplete="off" name="GUILD" placeholder="GUILD" maxlength="64" onkeyup="nempty_keyup('GUILD','guildchk')" onblur="nempty_keyup('GUILD','guildchk')"></input>
			<input type="text" id="PHONE" autocomplete="off" name="PHONE" placeholder="PHONE" onkeyup="tel_keyup()" onblur="tel_keyup()" maxlength="64" ></input>
			<input type="email" id="EMAIL" autocomplete="off" name="EMAIL" placeholder="EMAIL" onkeyup="em_keyup()" onblur="em_keyup()" maxlength="100"></input>
			<input type="file" style="display:none" id="FILE" name="FILE" onchange="filesizechk(this)" accept="image/*" ></input>
	
			<input type="text" id="fakeFile" value="" style="width:200px;" readonly="readonly"></input>
			<input type="button" value="UPLOAD" onclick="goFile()" style="width:90px;"></input>
			<input type="submit" value="Sign up" onclick="return CheckSubmit();"></input>
		</section>
	</div></div></div></div>
	</form>
	
</body>
</html>
EOF
  ;
